src/wtf8/mod.rs

```diff
@@ -23,17 +23,45 @@ pub struct Wtf8 {
 }
 
 impl Wtf8 {
+    /// Only safe if `bytes` contains valid WTF-8.
+    #[inline]
+    pub(crate) unsafe fn from_bytes_unchecked(bytes: &[u8]) -> &Self {
+        // Safety: the cast is sound because repr(transparent), matching the layout of [u8].
+        // As long as `bytes` is valid WTF-8, the type invariant is upheld.
+        &*(bytes as *const [u8] as *const Wtf8)
+    }
+    
+    /// Only safe if `bytes` contains valid WTF-8.
+    #[inline]
+    pub(crate) unsafe fn from_bytes_unchecked_mut(bytes: &mut [u8]) -> &mut Self {
+        // Safety: the cast is sound because repr(transparent), matching the layout of [u8].
+        // As long as `bytes` is valid WTF-8, the type invariant is upheld.
+        // The pointer is guaranteed to be safe for writes, as it comes from a mut ref.
+        &mut *(bytes as *mut [u8] as *mut Wtf8)
+    }
+    
     #[inline]
     pub(crate) fn bytes(&self) -> &[u8] {
         &self.bytes
     }
+    
+    #[inline]
+    pub(crate) fn bytes_mut<'a>(&'a mut self) -> &'a mut [u8] {

```

Comment:
`bytes_mut` must be unsafe: If the bytes are modified such that the string is no longer valid WTF-8, the result is unsound and may result in undefined behaviour.

See also: [`str::as_bytes_mut`](https://doc.rust-lang.org/std/primitive.str.html#method.as_bytes_mut) is unsafe.